### PR TITLE
Feature/update water simulation

### PR DESCRIPTION
**Pull request by AHABID**

### CODE REVIEW

**Comment by csweetapple** [comment]
A general comment - I'm a bit confused by the different entity types/names. I'm presuming 'Simulation' and 'Result' both different types of entity, as the folder structure and separate schemas etc. suggest. In which case;
* Why do they both have a required `type` of `Simulation` specified in their respective schemas?
* Why is what appears to be the results entity in the SimulationDataModel.jpg called `urn:ngsi-ld:Simulation:Results` rather than `urn:ngsi-ld:Results:name_of_results`? (I thought the name of the entity was given after `ngsi-ld:`)
* Would it be clearer to name the two entity types `SimulationScenario` and `SimulationResults`?

I'll add more comments on this inline.

**Comment by csweetapple** [comment]
I don't seem to be able to comment on the SimulationDataModel.jpg file directly - but the property/relationship names in this should be updated to match those in the schemas. E.g. `outcomes` changed to `outputParameters`.

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: `Result`, not `Simulation`.

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: This is the Simulation desctiption not the Result description

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: Result

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: typo

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: Result entity

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: Result entity.

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: Missing definition of `hasSimulation`.

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: Is it necessary to repeat `hasInputNetwork` in the `Result` entity when it can already be determined from the `Simulation` entity to which the `Result` is already connected via `hasSimulation` (I'm presuming `hasSimulation` will be a mandatory relationship).

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: Should this be updated to reflect the properties/relationships specified as mandatory in the spec?

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: typo in `Relationship`

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: typo in `Relationship`

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: typo in `Relationship`

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: THis shouldn't be here? This is from the `Simulation` entity?

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/doc/spec.md]: `targetURI`

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: This doesn't capture the required nesting of the properties (i.e. `targetURI` is a sub-relationship of the water attribute property) - should it?

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: Does the water attribute sub-property need to be captured somehow? (not sure how since it could be one of many possible attributes)

File: WaterNetworkManagement/Simulation/Simulation/README.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/README.md]: harmonised description of a simulation *scenario*? Could also change entity name to `SimulationScenario` to improve clarity. If doing this, also change the `hasSimulation` relationship in the Result entity to `hasSimulationScenario`, and potentially change the `Result` entity to `SimulationResult`.

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Result/schema.json]: I don't think `has` is the right verb to use for a relationship in this direction - I'd say a simulation scenario has a result rather than a result has a simulation scenario. A result is *from* a simulation scenario.

File: WaterNetworkManagement/Simulation/Simulation/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/doc/spec.md]: I thought there was an agreement to go for the generic `optionNames` and `optionValues` lists so that the Simulation data model wasn't tied to the requirements of EPANET?
If specifiying all the options individually then we need a lot more than just `duration`, `hydraulicTimeStep`, `flowUnits`, `headlossFormula` and `demandModel` to capture everything that might be required to fully define an EPANET simulation.

File: WaterNetworkManagement/Simulation/Result/doc/spec.md
Comment (AHABID) [reply WaterNetworkManagement/Simulation/Result/doc/spec.md]: It is not required so we can put it as optional

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (AHABID) [reply WaterNetworkManagement/Simulation/Result/schema.json]: In the $ref, they define a field called "additionnalProperties" which will allow us to add a sub-property or a sub-relationship. `targetURI` is required that's true, but I would prefer simple schema.

File: WaterNetworkManagement/Simulation/Simulation/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Simulation/doc/spec.md]: E.g.
* duration
* hydraulicTimeStep
* qualityTimeStep
* patternStep (global value - could potentially omit if mandatory in `Pattern` data model)
* patternStart (as for patternStep)
* reportStep (perhaps unimportant if results are extracted programatically and stored with the Result entity rather than the EPANET generated rpt file)
* reportStart (as for reportStep)
* ruleStep
* statistic
* trials
* accuracy
* tolerance
* EmitterExponent
* headError
* flowChange
* headlossFormula
* demandCharge
* specificGravity
* viscosity
* unbalanced
* checkFrequency
* maxCheck
* dampLimit
* diffusivity
* bulkOrder
* wallOrder
* tankOrder
* concentrationLimit
* qualityType
* chemicalName
* chemicalUnits
* traceNodeID

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (AHABID) [reply WaterNetworkManagement/Simulation/Result/schema.json]: Yes, this is done by the ` {
            "$ref": "https://raw.githubusercontent.com/smart-data-models/dataModel.WaterNetworkManagement/master/WaterNetworkManagement-schema.json"
        }`  ( line 12 of the schema file)
Maybe the next mission is to update the water management schema in order to support  all water attributes

File: WaterNetworkManagement/Simulation/Simulation/examples/example-normalized-ld.jsonld
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/examples/example-normalized-ld.jsonld]: This should be `monitoredNode` and `controlledLink` (not `targetURI`), as in "Operational Control 1".

File: WaterNetworkManagement/Simulation/Simulation/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/schema.json]: simulation *scenario*?

File: WaterNetworkManagement/Simulation/Simulation/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/schema.json]: `hasResult` is missing from the spec

File: WaterNetworkManagement/Simulation/Simulation/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/Simulation/schema.json]: Should capture all of the mandatory properties/relationships?

**Review by csweetapple — CHANGES_REQUESTED** [review]

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (AHABID) [reply WaterNetworkManagement/Simulation/Result/schema.json]: refSimulationScenario maybe

File: WaterNetworkManagement/Simulation/Simulation/doc/spec.md
Comment (AHABID) [reply WaterNetworkManagement/Simulation/Simulation/doc/spec.md]: I switched it to NGSI-LD properties for making queries such as "get the simulation accuracy" faisable.  I will add these properties to the schema, then maybe it will be good if you can add examples in the payload;

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Result/schema.json]: I prefer this.

**Comment by AHABID** [comment]
So, the Jpg figure is now updated with the current data model version. I have tried to check all your comments.
After validating this PR, I think that it  will be great if you can:
- make a check on the mondatory/optional fileds of spec.md
- add definition of the new  added properties in spec.md of SimulationScenario
- add some examples of the new added properties in the SimulationScenario exmaple.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md]: `SimulationScenario`, singular.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: `qualityTimeStep` missing from schema and spec

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:218]: This should be optional (only required if the quality type is `TRACE`)

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:133]: Values should be restricted to `NONE`, `CHEM`, `AGE` and `TRACE`.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: Descriptions for new properties to be added.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:21]: backticks missing from names of new properties.

File: WaterNetworkManagement/Simulation/Simulation/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Simulation/doc/spec.md]: `qualityTimeStep` still missing from schema and spec

File: WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld]: "urn:ngsi-ld:SimulationResult:01"

File: WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld:96]: Format of the the first example here is different to the others. Where have `applyToEntity` and `forAttribute` come from?

File: WaterNetworkManagement/Simulation/SimulationScenario/schema.json
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/schema.json]: See https://github.com/smart-data-models/incubated/pull/22#discussion_r516817112

File: WaterNetworkManagement/Simulation/Simulation/schema.json
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Simulation/schema.json]: `hasSimulationResult` now in spec.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md]: ok

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: ok

Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: I will add it at the same time as adding descriptions and examples.

Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: Just to avoid conflicts issues I'm reviwing your comments, I will commit these modifications. We merge the PR to master and then add the description and exmaples

File: WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld]: ok

File: WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/examples/example-normalized-ld.jsonld:96]: yes this was another proposition. 
updated

File: WaterNetworkManagement/Simulation/SimulationScenario/schema.json
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/schema.json]: ok

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:218]: Did not got it. 
Maybe if you can check the mondatory/option fileds, add descriptions and restrcitions to all new attributes

**Comment by AHABID** [comment]
@csweetapple I have added new commit. You can proceed on it (or on a new branch after merging this one).
Maybe you are more expert in adding descriptions and restrcitions about new properties in the spec.md file.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: OK

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:218]: I'll change this after your PRs merged.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:133]: I will update this when updating descriptions.

Comment (AHABID) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:133]: Ok Great! so I need your approval to merge this PR :)

File: WaterNetworkManagement/Simulation/Result/schema.json
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Result/schema.json]: Moving to new issue.

File: WaterNetworkManagement/Simulation/Simulation/schema.json
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/Simulation/schema.json]: Moving to new issue.

**Comment by csweetapple** [comment]
> 
> 
> So, the Jpg figure is now updated with the current data model version. I have tried to check all your comments.
> After validating this PR, I think that it will be great if you can:
> 
>     * make a check on the mondatory/optional fileds of spec.md
> 
>     * add definition of the new  added properties in spec.md of SimulationScenario
> 
>     * add some examples of the new added properties in the SimulationScenario exmaple.

Still a few changes needed in the figures I think.

In the WaterNetworkdataModel-Example.jpg:
* `urn:ngsi-ld:Network:SophiaWaterNetwork` should be `urn:ngsi-ld:WaterNetwork:...`?
* Pipes must start and end at a node (Junction, Reservoir or Tank) - so the `startsAt` relationships between the pipes and Pump:01 are not valid.
* `hasSubGraph` should be `hasSubNetwork`

In SimulationDataModel.jpg:
* `hasSimulation` needs to be changed to `refSimulationScenario`
* The `operationalControl` properties and relationships don't match the schema/spec. `setting` is missing from Operational Control 2 and `targetURI` needs to be replaced with `controlledLink` and `monitoredNode` relationships. In Operational Control 1, the `controlled/monitored` relationship needs to be replaced with `controlledLink` and `monitoredNode`, and `triggerLevel` is a property of the operational control, not the tank.
* `inputParameters` should be `inputParameter` according to the spec.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [inline WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md]: should be reference to an entity of type `SimulationResult`

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:16]: Moving to new issue.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:133]: Moving to new issue.

**Comment by csweetapple** [comment]
@AHABID I've marked most of the comments as resolved - just a couple outstanding, plus some changes needed in the diagrams, then I'll merge and add descriptions and restrictions etc. in a new branch.

**Comment by AHABID** [comment]
Yes done now

**Comment by csweetapple** [comment]
Still some more changes needed on the diagrams...

In the WaterNetworkdataModel-Example.jpg:

* Pipes must start and end at a node (Junction, Reservoir or Tank) - so the `startsAt` relationships between the pipes and Pump:01 are not valid.
 
 
In SimulationDataModel.jpg:
* `hasSimulationScenario` needs to be changed to `refSimulationScenario`
* In Operational Control 1:
  * the `monitoredNode` relationship needs to connect to a node (e.g. `urn:ngsi-ld:Tank:T1`.
  * the `controlledLink` relationship needs to connect to a link (e.g. `urn:ngsi-ld:Pump:P1`.
* In Operational Control 2:
  * the `monitoredNode` relationship needs to connect to a node (e.g. `urn:ngsi-ld:Tank:T2`.
* the second relationship should be `controlledLink` not `controlled/monitored`
  * the `controlledLink` relationship needs to connect to a link (e.g. `urn:ngsi-ld:Pump:P2`.

File: WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md
Comment (csweetapple) [reply WaterNetworkManagement/Simulation/SimulationScenario/doc/spec.md:21]: This wasn't resolved.

**Comment by AHABID** [comment]
Maybe if you can change them directly on the graphml file (usin yed)

**Pull request merged**
